### PR TITLE
dialects: (dmp) A bunch of fixes in the lowering

### PR DESCRIPTION
**Pull request by AntonLydike**

### CODE REVIEW

File: xdsl/transforms/lower_mpi.py [hunk 1]
```diff
@@ -211,7 +214,7 @@ def _emit_memref_counts(
         if not all(dim.value.data >= 0 for dim in ssa_val.typ.shape.data):
             raise RuntimeError("MPI lowering does not support unknown-size memrefs!")
 
-        size = sum(dim.value.data for dim in ssa_val.typ.shape.data)
+        size = prod(dim.value.data for dim in ssa_val.typ.shape.data)
```
Comment (superlopuh) [inline xdsl/transforms/lower_mpi.py:217]: seems like a useful helper to add on the type, potentially?

File: xdsl/transforms/experimental/dmp/stencil_global_to_local.py [hunk 2]
```diff
@@ -235,100 +236,58 @@ def generate_memcpy(
...
-                load = memref.Load.get(source, [x, y])
-                store = memref.Store.get(load, dest, [linearized_idx])
-            yield from (linearized_idx, load, store)
-        yield scf.Yield.get()
-
-    def loop_body_with_for(i: SSAValue) -> Iterable[Operation]:
+    cst0 = arith.Constant.from_int_and_width(0, idx)
+    cst1 = arith.Constant.from_int_and_width(1, idx)
+
+    @Builder.implicit_region([idx, idx])
```
Comment (superlopuh) [inline xdsl/transforms/experimental/dmp/stencil_global_to_local.py:253]: yay! clients!

**Review by superlopuh — APPROVED** [review]

File: xdsl/transforms/experimental/dmp/stencil_global_to_local.py (hunk 2 above)
Comment (AntonLydike) [reply xdsl/transforms/experimental/dmp/stencil_global_to_local.py:258]: It's so wonderful to use! Thanks for adding it!

File: xdsl/transforms/lower_mpi.py (hunk 1 above)
Comment (AntonLydike) [reply xdsl/transforms/lower_mpi.py:217]: Absolutely!

**Review by georgebisbas — APPROVED** [review]

**Pull request merged**